r/tests/testthat/test-dplyr-string-functions.R

```diff
@@ -719,6 +719,40 @@ test_that("errors in strptime", {
   )
 })
 
+test_that("strftime", {
+  skip_on_os("windows") # https://issues.apache.org/jira/browse/ARROW-13168
+
+  # TODO: consider reevaluating this workaround after ARROW-12980
+  withr::local_timezone("UTC")
+  times <- tibble(x = c(lubridate::ymd_hms("2018-10-07 19:04:05"), NA))
+
+  formats = c("%a", "%A", "%w", "%d", "%b", "%B", "%m", "%y", "%Y", "%H",
+             "%I", "%p", "%M", "%z", "%Z", "%j", "%U", "%W", "%c", "%x",
+             "%X", "%%", "%G", "%V", "%u")
+
+  for (format in formats) {
+    expect_dplyr_equal(
+      input %>%
+        mutate(x = strftime(x, format = format)) %>%
+        collect(),
+      times
+    )
+
+    expect_dplyr_equal(
+      input %>%
+        mutate(x = strftime(x, format = format, usetz = TRUE)) %>%
+        collect(),
+      times
+    )
+
+    x <- Expression$field_ref("x")
+    expect_error(
+      nse_funcs$strftime(x, format = format, tz="Mars/Mariner_Valley"),
+      "tz argument not supported by Arrow"
+    )
+  }
+})
```

Comment:
This is looking a lot more along the right lines, but I reckon that this could be refactored further to be more efficient, and easier to diagnose what's happening if any of the tests fail.  

Please would you refactor so that you remove the `for` loop, and have `format` as a column in the tibble (you won't need to change the `mutate` statement), so we just run each of those `expect_*` functions once?

Reply:
You mean like this?
```R
  df <- tibble(
    x = lubridate::ymd_hms("2019-02-03 12:34:56"),
    y = c("%a", "%A", "%w", "%d", "%b", "%B", "%m", "%y", "%Y", "%H",
          "%I", "%p", "%M", "%z", "%Z", "%j", "%U", "%W", "%c", "%x",
          "%X", "%%", "%G", "%V", "%u")
  )

  expect_dplyr_equal(
    input %>%
      mutate(x = strftime(x, format = y)) %>%
      collect(),
    df
  )
```

It's failing with:
```bash
 Failure (test-dplyr-string-functions.R:735:3): strftime
`via_batch <- rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(input = record_batch(tbl))))` generated warnings:
* Expression strftime(x, y) not supported in Arrow; pulling data into R
Backtrace:
 1. expect_dplyr_equal(...) test-dplyr-string-functions.R:735:2
 2. testthat::expect_warning(...) helper-expectation.R:101:4
```

My R is not enough. I'll try again tomorrow.

Reply:
Apologies, my mistake; I'd presumed that we can use a column of a table as something to pass into the `options` param, but it appears we can't. 

Perhaps the solution here is writing multiple separate tests, though I don't see why you couldn't combine multiple format markers in one test to have a smaller subset of tests.

Reply:
How about something like this?  It's not complete - the final two tests are still causing problems, but it's a starting point.

```suggestion
df <- tibble(x = c(lubridate::ymd_hms("2019-02-03 12:34:56"),  NA))

expect_dplyr_equal(
  input %>%
    mutate(x = strftime(x, format = "%a %A %w %d %b")) %>%
    collect(),
  df
)
 
expect_dplyr_equal(
  input %>%
    mutate(x = strftime(x, format = "%B %m %y %Y %H")) %>%
    collect(),
  df
)

expect_dplyr_equal(
  input %>%
    mutate(x = strftime(x, format = "%I %p %M %j")) %>%
    collect(),
  df
)

expect_dplyr_equal(
  input %>%
    mutate(x = strftime(x, format = "%U %W %c %x")) %>%
    collect(),
  df
)

expect_dplyr_equal(
  input %>%
    mutate(x = strftime(x, format = "%X %% %G %V %u")) %>%
    collect(),
  df
)

# Needs updating to deal with the difference in precision
expect_dplyr_equal(
  input %>%
    mutate(x = strftime(x, format = "%S")) %>%
    collect(),
  df
)

# we should do something around documenting or dealing with this difference too
withr::with_timezone("Pacific/Marquesas",
 expect_dplyr_equal(
    input %>%
      mutate(x = strftime(x, format = "%z %Z ")) %>%
      collect(),
    df
  )
)

})
```

Reply:
Oh yeah indeed. We can also just do a single string with all the flags :)